### ISSUE TITLE
Property prediction

### ISSUE THREAD

**Issue by dribnet** [maxhodak/keras-molecules#11]
This codebase works well for me and I'm able to replicate the current results. Having worked a good bit with other latent spaces, I'm curious to find out what other operations the latent space of this model might support.  Specifically, I suspect that the latent space could also support analogies and attribute vectors, but unfortunately I'm not familiar with chemistry datasets and smile strings.

Would anyone be interested in helping me build a labelled dataset of molecules that includes binary attributes and then investigating the results of applying attribute vectors? An example structure of the dataset would be:

| smile string  | Polar | Toxic |  Flammable |  Positive Oxidation State |
| ------------- | ------------- |------------- |------------- |------------- |
| CN1CCC[C@H]1c2cccnc2  | True  | False | False | True |
| O=C1Oc2ccccc2c3ccccc13  | False | True  | True  | False |
| ...  | |  |  | |

Generally, these datasets can be useful even if they are much smaller than the training dataset - say dozens to hundreds of rows. Ideally, the chosen attributes would be those that could serve as unambiguous labels and operators. For example, pretend the following is true:

```
Carbon dioxide is a polar molecule.
The equivalent to carbon dioxide without polarity is carbon monoxide.
```

Then this would be a great attribute because it follows the formula:

```
Molecule X has (doesn't have) attribute Y
The equivalent of Molecule X with (without) attribute Y is Z
```

I don't know enough chemistry to know if there are even such attributes for subsets of molecules. But if there are, then a small dataset of molecules with and without attribute Y would be sufficient to see if Z could be inferred from this model given X.

**Comment by hsiaoyi0504** [comment]
According to original paper, properties are labelled by calculation of LogP from [RDKit](http://www.rdkit.org/). I somewhat know how to use it, so if it helps you?

**Comment by hsiaoyi0504** [comment]
And to further re-implementation of original models included in original paper.
I think this repo will be also useful
https://github.com/fmfn/BayesianOptimization

**Comment by dribnet** [comment]
With a tip from @rpadams I was able to locate a solubility dateset (J.S. Delaney ESOL: Estimating aqueous solubility directly from molecular structure. J. Chem. Inform. and Comp. Sci., 44(3):1000–1005, 2004.) made available by @duvenaud [in the neural-fingerprint repo](https://github.com/HIPS/neural-fingerprint/tree/master/data/2015-05-24-delaney). With light modifications to this codebase (pull #25) I was able to encode this dataset into the latent space of the provided `model_500k.h5`, and then used my [plat library](https://github.com/dribnet/plat/) I was able to encode a "solubility vector".

I've attached this 'solubility vector' in json form if anyone wants to experiment with it. The idea is that:

1) any smiles molecule could be encoded using the `model_500k.h5` into latent space
2) The 'solubility vector' could be applied to move in latent space toward more solubility. Note that it can be applied with a scalar, include a negative scalar to make it less soluble. Generally the useful range of the scalar would be [-1, 2]
3) This vector could be decoded back into a molecule and interpreted

(3) is the difficult part for me - so I'm not sure if this will work or not. But I did do some testing on hold-out data from the delaney dataset that hinted it might be working. Also - this is based on a pretty small dataset and it might contain other bias in the labelings.

[solubility_vector.json.zip](https://github.com/maxhodak/keras-molecules/files/586798/solubility_vector.json.zip)

**Comment by maxhodak** [comment]
cc @dakoner re: our recent emails on zinc properties

**Comment by maxhodak** [comment]
I've overhauled the dataset downloading scripts to start getting into this problem. ZINC12 is an ok reasonably large dataset that comes with a bunch of nice properties for us to explore. To get it, you can now do:

`python download_dataset.py --dataset zinc12`

This also works with `--dataset chembl22` but that doesn't come with any properties, just SMILES.

You'll need to re-run the preprocessing script with a `--property_column`:

```
python preprocess.py
    --smiles_column SMILES
    --property_column LogP
    data/zinc12.h5 data/processed.h5
```

That will produce a 7.7G data file with the property data in `property_train` and `property_test` fields corresponding to the entries in `data_train` and `data_test`.

**Comment by hsiaoyi0504** [comment]
I think you mean `--dataset zinc12` ?(A typo)

**Comment by maxhodak** [comment]
Yes, edited my comment to match.

**Comment by hsiaoyi0504** [comment]
Another thing I think need to be mention is that ZINC database is shipped with `xlogP`.
I don't know it is equivalent to logP calculated from `RDKit`'s `MolLogP`

**Comment by maxhodak** [comment]
The column is named `LogP`... it's not clear if it's measured or calculated, but I would have expected it to be `cLogP` if it was calculated. I know a big point of ZINC is to include orderable compounds which also makes me think there's a good chance they're measured. On the other hand, that's a ton of compounds to have measured data for...

**Comment by maxhodak** [comment]
The full list of columns in the ZINC12 file is:

```
- Charge
- Desolv_apolar
- Desolv_polar
- HBA
- HBD
- LogP
- MWT
- NRB
- SMILES
- ZINC_ID
- tPSA
```

**Comment by hsiaoyi0504** [comment]
As far as I know from [here](http://wiki.bkslab.org/index.php/ZINC_subsets). It's derived from calculation.

**Comment by dakoner** [comment]
I suggest loading all the properties en masse into the training/test data.
One of the suggestions we got from Geoff Hinton and George Dahl was that
forcing the learning to predict more things simultaneously would make a
better model (less overfitting?)

**Comment by hsiaoyi0504** [comment]
I agree the suggestions in your comment, but the original work we want to reimplement here is just to find a continuous representation of chemical derived from current string representation. And back to original scenario, namely, drug design, we usually can't know all properties of the drugs, so maybe you suggested here is somewhat unreasonable? The original work done in the paper is simply train a autoencoder NN and  another GP model. As far as I know, GP models don't have such multiple output characteristic as NN.

**Comment by maxhodak** [comment]
I think one of the decisions to be made is what the ambition of this repo is: are we just reproducing the one paper? Do we want to expand upon it to some other functional aim?

Personally I'm for going further as there are lots of interesting ideas to explore here and not many venues like this where you can just check out the code and easily get something running. I also like the little proto-community we have here and I don't know how well it would work to move everyone to a different repo.

**Comment by hsiaoyi0504** [comment]
I agree what you point out, and I think go further is better, but I think before this, the repetition of original model is needed. Currently, I think the GP part of original paper is what you didn't achieve.
